### PR TITLE
Migrate away from Eq1 / Ord1 / Show1

### LINKED ISSUE #27: Migrate away from Eq1/Ord1/Show1
Continuing on https://github.com/haskell/core-libraries-committee/issues/190#issuecomment-1657199739, I propose to migrate from

```haskell
instance Eq1 f => Eq (Fix f)
instance Ord1 f => Ord (Fix f)
instance Show1 f => Show (Fix f)
```
to 

```haskell
instance Eq (f (Fix f)) => Eq (Fix f)
instance Ord (f (Fix f)) => Ord (Fix f)
instance Show (f (Fix f)) => Show (Fix f)
```

or if you fancy `QuantifiedContraints`

```haskell
instance (forall a. Eq a => Eq (f a)) => Eq (Fix f)
instance (forall a. Ord a => Ord (f a)) => Ord (Fix f)
instance (forall a. Show a => Show (f a)) => Show (Fix f)
```

The motivation is that I recently migrated a codebase from a homegrown `newtype Fix` with `instance Ord (f (Fix f)) => Ord (Fix f)` to `data-fix` and hit performance issues. They were a combination of several factors:

* I used [`Data.Functor.Classes.Generic`](https://hackage.haskell.org/package/transformers-compat-0.7.2/docs/Data-Functor-Classes-Generic.html) to derive `Ord1` instance generically. It appeared that [`from`](https://hackage.haskell.org/package/base-4.19.1.0/docs/GHC-Generics.html#v:from) and [`to`](https://hackage.haskell.org/package/base-4.19.1.0/docs/GHC-Generics.html#v:to) survived in Core, despite pretty aggressive optimization options. This is admittedly not a fault of `data-fix`, but a general issue with usability of `Eq1` / `Ord1` / `Show1`. I had to write all instances manually.

* [`class Ord1`](https://hackage.haskell.org/package/base-4.19.1.0/docs/Data-Functor-Classes.html#t:Ord1) contains only [`liftCompare`](https://hackage.haskell.org/package/base-4.19.1.0/docs/Data-Functor-Classes.html#v:liftCompare), but no `liftLT`, `liftLE`, `liftGT`, `liftGE`, so it has no access to `<`, `<=`, `>=`, `>` of the base functor. This implies that `<` on `Fix f` also necessarily goes through `compare` and takes a performance hit.

* It does not seem that `liftEq` / `liftCompare` can be optimized away unless the base functor `f` is completely monomorphic and everything is forced to inline. E. g., `(==)` specialised to `Fix (ListF Int)` looks reasonable, but `(==)` for `Fix (ListF a)` goes through `liftEq`. The Core for 

    ```haskell
    instance Eq (f (Fix f)) => Eq (Fix f) where
      (==) = coerce ((==) @(f (Fix f)))
    ```

    looks more reasonable to me.

----

I understand that the proposed change is a breaking one. It is however not very pronounced: in recent `base` we have `class (forall a. Eq f a) => Eq1 f`, so the majority of users will not even notice. It's only if you require `Eq (Fix f)` as a constraint, but use `liftEq` inside, which will explode, because `Eq1 f` is no longer guaranteed. I think, however, that GHC would yell if encounters `Eq (Fix f)` constraint, suggesting to replace it with `Eq1 f`.

@phadej what do you think?

### PR DESCRIPTION
**Pull request by Bodigrim**
Closes #27. 

Tested with 
```
packages: .
packages: dhall-1.42.1
packages: aeson-2.2.1.0
packages: hnix-0.17.0
packages: sexp-grammar-2.3.4.2
packages: recursion-schemes-5.2.2.5
packages: quickcheck-instances-0.3.30
packages: liquidhaskell-boot-0.9.8.1

allow-newer: hnix:aeson, hnix:free, hnix:bytestring, hnix-store-core:algebraic-graphs
```

I'm on aarch64 so unfortunately cannot test locally with old GHCs. 

Really old GHC <= 7.6 fail in CI. If we must retain compatibility with them, I can revert changes with regards to `Typeable1`.

### CODE REVIEW

**Comment by Bodigrim** [comment]
@phadej I got access to a x86_64 machine and tested the patch with GHC 8.4 and 8.6. No breakage in downstream packages detected.

**Comment by Bodigrim** [comment]
@phadej anything else you would like me to check here?

**Comment by phadej** [comment]
Everything looks fine.

I'm waiting for GHC-9.10 release; as this is still a breaking change, I'd like to make a 9.10 compatibility work first, and only then make a major release.

File: src/Data/Fix.hs [hunk 1]
```diff
@@ -266,18 +270,22 @@ fixDataType = mkDataType "Data.Functor.Foldable.Fix" [fixConstr]
 -- | Least fixed point. Efficient folding.
 newtype Mu f = Mu { unMu :: forall a. (f a -> a) -> a }
 
-instance (Functor f, Eq1 f) => Eq (Mu f) where
+instance (Functor f, Eq (f (Fix f))) => Eq (Mu f) where
     (==) = (==) `on` foldMu Fix
 
-instance (Functor f, Ord1 f) => Ord (Mu f) where
+instance (Functor f, Ord (f (Fix f))) => Ord (Mu f) where
```
Comment (phadej) [inline src/Data/Fix.hs:276]: Now thinking more about it. I'm not happy about this change. It exposes too much of implementation detail to the surface. `Ord1 f` is asking for more, but doesn't tell anyone on the outside that `Fix` is used internally.

I'm not particularly happy with this. So I'm sorry, but we'll need to rethink this change.

Changing `Fix` interface is one option, but I don't like that either, as then `Fix` and `Mu/Nu` parts will look different.

**Review by phadej — CHANGES_REQUESTED** [review]

**Comment by phadej** [comment]
One option, which is not breaking, and may fix part of #77 is to use `forall a. Ord a => Ord (f a)` implied by `Ord1 f` in the implementation, when such implication exists (i.e. more recent `base`).

**Comment by Bodigrim** [comment]
> One option, which is not breaking, and may fix part of #27 is to use `forall a. Ord a => Ord (f a)` implied by `Ord1 f` in the implementation, when such implication exists (i.e. more recent `base`).

Works for me. Do you want it only for `Mu` and `Nu` or for `Fix` also? Shall I update the PR?

**Comment by Bodigrim** [comment]
Closing in favor of #29.

**Pull request closed without merging**